File: Leetcode/206. Reverse Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode *reverseList(struct ListNode *head)
{
    struct ListNode *pre = NULL;
    struct ListNode *cur = head;
    struct ListNode *next = NULL;
    while (cur != NULL)
    {
        next = cur->next;

        cur->next = pre;
        pre = cur;
        cur = next;
    }
    return pre;
}

File: Leetcode/206. Reverse Linked List.java
class Solution {
    public static class ListNode {
        int val;
        ListNode next = null;

        ListNode(int val) {
            this.val = val;
        }
    }

    public ListNode reverseList(ListNode head) {
        ListNode dummy = new ListNode(0);
        ListNode cur = null, next = null;
        while (head != null) {
            next = head.next;
            dummy.next = head;
            head.next = cur;
            cur = head;
            head = next;
        }
        return dummy.next;
    }
}